### ISSUE TITLE
Performance and stability targets

### ISSUE THREAD

**Issue by zhenyasav** [dxos/dxos#3862]
## Goals

The following goals are our target for the Strangeloop epoch:

- Able to effectively dogfood Composer within the team.
- Able to effectively market Composer and DXOS in general to build collaborative apps.
- Able to measure and monitor stability and performance via automated tests.
- Able to define a baseline against which we can prioritize and consider optimizations.

### Details

- Users
  - 5 concurrent devices per profile
- Spaces
  - 10 active spaces
  - 10 online members per space (with multiple devices)
  - 20 composer documents per space
  - 500 objects per space
  - 5 simultaneous invitations
  - Ability for users to generate epochs (either manually, or via agents)
- Metrics
  - ~1s cold-start per space (staggered)
  - ~10s full space replication
  - epochs triggered every 5,000 mutations

## Issues

- [x] Configure hypercore replication
- [x] Change to "fast" WebFS API and/or batch writes
- [ ] Minimize network requests during replication (currently 1 request per block)
- [x] Batch mutations
- [x] Merge mutations (esp. YJS)
- [ ] Trigger epoch before invitation
- [ ] Save local snapshot before exit
- [ ] #3895
- [x] Connection limiting
- [ ] Research how libp2p makes webrtc handshake faster
- [ ] Protobuf code loading optimization
- [ ] Metadata save intervals tuning
- [ ] Application mitigations
  - [ ] Set limits
  - [ ] Warnings
  - [ ] Error logging (Sentry)

## Reference

- Discord #protocols (Benchmark thread)
- https://github.com/dxos/dxos/issues/3802

**Comment by nferch** [comment]
@zhenyasav is there some reference for how we define "cold-start"?

**Comment by richburdon** [comment]
TL;DR: this is a north start issue that sets the minbar for platform stability and performance that we need to address before the conference in order to have credibility.
[8:58 PM]Rich: there are many factors at play, and there are multiple issues currently being worked on that relate to this, but i'll call out FOUR specific initiatives:

connection stability: the ability to tolerate multiple concurrent invitations AND multiple concurrent swarm connections -- via connection limiting.

replication performance: the expectation to replicate spaces (post invitation) within seconds (10s max); this relates to hypercore optimizations (reducing the # of roundtrip messages and how we use the browser file APIs)

cold-start performance: the expectation to start using the app within seconds (per space); this relates to the browser file APIS and to merging mutations (e.g., time-buffering text document edits into single mutations).

epochs/snapshots: the ability for users to trigger epochs (manually or via agents); and for the system to automatically trigger epochs/snapshots on creating an invitation and/or before exiting.
